doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+
+The only requirment is that for FATAL errors the `RAISE_ERROR` invocation shall never return and lead to termination.
+
+It is also possible to provide no `error_code` at all, which leads t a generic FATAL error:
+
+`RAISE_ERROR()`.
+
+A further conisderation is allowing custom messages (in principle not necessary if we allow custom handler functions)
+which are delegated to the handling code. The handling code can then decide to log them or otherwise incorporate
+it into the reporting code.
+
+`RAISE_ERROR(error_code, FATAL, "Mutex corrupted")`.
+
+The problem of various optional arguments is to define a proper argument order and, if macros are used,
+the requirement of variadic macros. It is generally not possible to leave out arguments in the middle
+without defining an entirely different macro or function.
+The macro wrappers should be very thin and directy delegate to a template function that enables
+the compiler to optimize the code.
+
+### Non-recoverable Error Handling API
+
+Non recoverable errors could be raised with the general API as well as
+
+`RAISE_ERROR(error_code, FATAL, optional_message)`
+
+This assumes we have checked for some error `condition` first. We can incorporate the check in another macro
+and arrive at:
+
+`ASSERT(!condition, error_code, optional_message)`
+
+This is mainly syntactic sugar to increase readability and state intent in a more concise way.
+We can use both notations. Notice that this always operates at FATAL level and must be always active
+since there is no notion how we can continue normally from there (otherwise it would not be a FATAL error).
+
+Again optional arguments lead to a more involved implementation. Providing an optional handler is even more useful here
+(as the check is happens in the macro) and can look like this:
+
+`ASSERT(!condition, error_code, handler, optional_handler_args)`
+
+If the handler is not invocable with provided arguments (if any) then the compilation will fail.
+Unless we define another macro for the default error code, `error_code` must be provided.
+It is not advisable to manage this in the handler for consistency and to allow search for the error codes.
+
+### Generalization of ASSERT to Recoverable Errors
+
+If we want to support recoverable errors with a similar syntax to `ASSERT` we can define
+
+`EXPECT(condition, error_code, error_level, handler, optional_handler_args)`
+
+This is similar to Goofle Test terminology and basically states that we expect some `condition` to be true
+but do not necessarily terminate if it is not (depending on the `error_level`).
+If it is not true we raise an error of the required level.
+
+Hence
+`EXPECT(condition, error_code, FATAL, handler, optional_handler_args)`
+
+is equivalent to
+
+`ASSERT(condition, error_code, handler, optional_handler_args)`.
+
+If we do not need to call a handler and just provide some error message we can define
+
+`EXPECT(condition, error_code, error_level, error_message)`
```

Comment:
It is probably best to design it in a way that allows passing an optional error message with `operator<<` instead.